### PR TITLE
Issue 2881 - AgentAutoUpgrade: Add nmp check CLIs

### PR DESCRIPTION
**Pull request by Polber**
Signed-off-by: Jeff Kinard <jeff@thekinards.com>

### CODE REVIEW

File: cli/exchange/nmp.go [hunk 1]
```diff
@@ -155,3 +170,87 @@ func NMPRemove(org, credToUse, nmpName string, force bool) {
...
+				compatibleNodes = append(compatibleNodes, nodeNameEx)
+			}
+		} else {
+			// list policy
+			var nodePol exchange.ExchangeNodePolicy
+			_, nodeName := cliutils.TrimOrg(org, nodeNameEx)
+			cliutils.ExchangeGet("Exchange", cliutils.GetExchangeUrl(), "orgs/"+nmpOrg+"/nodes"+cliutils.AddSlash(nodeName)+"/policy", cliutils.OrgAndCreds(org, credToUse), []int{200, 404}, &nodePol)
+			nodePol_nm := nodePol.GetManagementPolicy()
+
+			if err := nodePol.Constraints.IsSatisfiedBy(nodePol_nm.Properties); err != nil {
```
Comment (linggao) [inline cli/exchange/nmp.go:246]: should use `nodePol.GetManagementPolicy().Constraints`

File: cli/exchange/nmp.go [hunk 2]
```diff
@@ -155,3 +170,87 @@ func NMPRemove(org, credToUse, nmpName string, force bool) {
...
+				continue
+			} else if err = nodePol_nm.Constraints.IsSatisfiedBy(nodePol.Properties); err != nil {
```
Comment (linggao) [inline cli/exchange/nmp.go:248]: should use `nodePol.GetManagementPolicy(). Properties `

File: cli/exchange/nmp.go [hunk 3]
```diff
@@ -155,3 +170,87 @@ func NMPRemove(org, credToUse, nmpName string, force bool) {
...
+func determineCompatibleNodes(org, credToUse, nmpName string, nmpPolicy exchangecommon.ExchangeNodeManagementPolicy) []string {
+	var nmpOrg string
+	nmpOrg, nmpName = cliutils.TrimOrg(org, nmpName)
+	
+	// get node(s) name(s) from the Exchange
+	var resp ExchangeNodes
+	cliutils.ExchangeGet("Exchange", cliutils.GetExchangeUrl(), "orgs/"+nmpOrg+"/nodes", cliutils.OrgAndCreds(org, credToUse), []int{200, 404}, &resp)
+
+	compatibleNodes := []string{}
+	for nodeNameEx, node := range resp.Nodes {
```
Comment (linggao) [inline cli/exchange/nmp.go:234]: Add a check to make sure that the node is registered.  If `node.PublickKey != ""`, then it is registered.

File: cli/exchange/node.go [hunk 4]
```diff
@@ -656,6 +656,51 @@ func NodeListStatus(org string, credToUse string, node string) {
 
 }
 
+func NodeListNMPs(org, credToUse, nodeName string, all bool) {
+	msgPrinter := i18n.GetMessagePrinter()
+
+	cliutils.SetWhetherUsingApiKey(credToUse)
+	var nodeOrg string
+	nodeOrg, nodeName = cliutils.TrimOrg(org, nodeName)
+
```
Comment (linggao) [inline cli/exchange/node.go:665]: Make sure the node is registered. If not, the nmp list is empty. To check if the node is registered, get the node resource from the exchange, make sure `PublicKey` is not empty.

File: cli/exchange/nmp.go (hunk 1 above)
Comment (Polber) [reply cli/exchange/nmp.go:246]: I'm not sure I understand this change, but if it is for clarity, I am updating the variable names to nodePolicy and nodeManagementPolicy for better distinction.

Comment (linggao) [reply cli/exchange/nmp.go]: nodePol_nm is the management part of the node policy, it should be verified against nmpPolicy instead of nodePol.

**Review by linggao — APPROVED** [review]

**Pull request merged**